### PR TITLE
Factory client hack

### PR DESCRIPTION
**Pull request by AntoineDao**
Proposing a client based on factory methods to make the code drier. Also means we can now instantiate [dataclass](https://docs.python.org/3/library/dataclasses.html) instances as a return from api calls.

The calls follow this style:
```python
from pyspeckle import SpeckleClient

client = SpeckleClient()

client.login(email='foo@bar.com', password='IeaTaVOCadoForbrEaKFasT')

projects = client.projects.list()

project = projects[0]

project.name = 'ItsAllGoneHorriblyWrong'

client.projects.update(project.id, project.dict())

client.projects.delete(project.id)
```

I can write some more extensive docs if we want to, but let's first see if this is useful and then I'll write them up proper :smile:

### CODE REVIEW

**Comment by AntoineDao** [comment]
Fixes #3 and #2

**Comment by tsvilans** [comment]
To list streams, we need to override the `list()` method to use `path="?omit=objects"` instead of `path="/"`.
@didimitrie I'm not sure why this is needed. If it's just `/` then the server returns an error.

**Comment by tsvilans** [comment]
Calling `client.objects.get(id)` returns `AttributeError: 'NoneType' object has no attribute 'parse_obj'`.

This happens in 
```
  File "\speckle\base\resource.py", line 101, in _parse_response
    return self.schema.parse_obj(response)
```

I can't find where `Resource.schema` gets set or initialized...

**Comment by tsvilans** [comment]
> I think we can rewrite the `make_request` method to accept `kwargs` as URL query parameters or pass them in the data argument (if it's a get request then dump data as query params).

Yeah, that sounds good. `kwargs` might be nice for stacking query filters and such.

**Comment by tsvilans** [comment]
Ok, I'm inclined to go ahead and merge, as now I've got it to generally work as expected, and it's not breaking anything from the older version.

The schema model is cool, it just needs to be implemented for all objects (tried it with a Mesh and it just ignores `vertices`, `faces`, etc. because they are not defined in the `SpeckleObject` schema). So I guess there will be some sort of dispatching or something.

@didimitrie and Antoine, any last thoughts or comments?

File: speckle/resources/objects.py [hunk 1]
```diff
@@ -46,6 +46,8 @@ def __init__(self, session, basepath):
                 'method': 'PUT'
             }
         })
+           
+        self.schema = SpeckleObject
```
Comment (AntoineDao) [inline speckle/resources/objects.py:50]: Good catch! Should've definitely not PR'd this at midnight :laughing: Got a bit too excited.

**Comment by AntoineDao** [comment]
Yeah I was thinking we could add more complex data classes for speckle objects in another PR to keep this one manageable. Pydantic will help creating different schemas depending to the `type` of the object. We can also add some fancy methods to those classes once they are set up (like area from polygon objects or line length and what not).

Other than that yeah I reckon it's good to merge :+1:

**Pull request merged**